torch/distributed/_shard/sharded_tensor/_ops/matrix_ops.py

```diff
@@ -0,0 +1,76 @@
+import torch
+import torch.distributed as dist
+from torch.distributed._shard.sharded_tensor import (
+    sharded_op_impl,
+    ShardedTensor,
+)
+
+@sharded_op_impl(torch.bmm)
+def sharded_bmm(types, args=(), kwargs=None, pg=None):
+    """
+    Handles ``__torch_function__`` dispatch for the sharded_bmm op.
+    """
+    input = args[0]
+    mat2 = args[1]
+
+    # Validate types
+    if not isinstance(input, ShardedTensor) or not isinstance(mat2, ShardedTensor):
+        raise TypeError("input or mat2 needs to be a ShardedTensor")
+    if input.dim() != 3 or mat2.dim() != 3:
+        raise TypeError("input or mat2 needs to be a 3D ShardedTensor")
+    local_tensor = torch.bmm(input.local_tensor(), mat2.local_tensor())
+    new_st_size = list(input.size())  # type: ignore[arg-type]
+    new_sharding_dim = input.sharding_spec().dim  # type: ignore[attr-defined]
+    mat2_sharding_dim = mat2.sharding_spec().dim  # type: ignore[attr-defined]
+    new_st_size[-1] = mat2.size()[-1]  # type: ignore[index]
+    if new_sharding_dim == -1 or new_sharding_dim == (input.dim() - 1):
+        if mat2_sharding_dim != -1 and mat2_sharding_dim != (mat2.dim() - 1):
+            new_st_size[-1] *= dist.get_world_size(pg)
```

Comment:
Sure, so the reason for this logic is that if the input is sharded by -1 and the mat2 is sharded in other dimensions (for example, 0 or 1). If we will want to shard it by -1, we then need to multiply the world_size because new_st_size[-1] now is the local tensor size.

Of course, there are many corner cases we need to add down the road.

Reply:
Resolved offline.